Python/compile.c

```diff
@@ -1949,6 +1949,25 @@ compiler_default_arguments(struct compiler *c, arguments_ty args)
     return funcflags;
 }
 
+static int corrected_firstlineno(struct compiler *c, stmt_ty s,
+                                 asdl_seq * decos)
+{
+        /* To keep the ability to get the relevant source of a decorated item
+        using inspect.getsource, we need to keep the first line number
+        of the code object as the first line number of the first decorator.
+        This used to be done via the AST, but it is better to hide this
+        internally.
+        */
+        if (asdl_seq_LEN(decos) > 0) {
+                expr_ty first_decorator = asdl_seq_GET(decos, 0);
+                c->u->u_firstlineno = first_decorator->lineno;
```

Comment:
Seems `c->u` still refers to the outer scope.

Add a regression test for these cases:
```python
def f():
    @deco
    def g(): pass
```
```python
class A:
    @deco
    def m(self): pass
```

Reply:
>Seems c->u still refers to the outer scope.

I'm probably missing something, but the generated AST is:
```
('Module',
 [('ClassDef',
   (1, 0),
   'A',
   [],
   [],
   [('FunctionDef',
     (3, 4),
     'm',
     ('arguments', [('arg', (3, 10), 'self', None)], None, [], [], None, []),
     [('Pass', (3, 17))],
     [('Name', (2, 5), 'deco', ('Load',))],
     None,
     None)],
   [],
   None)],
 None)
```
which seems correct. Perhaps I am misunderstanding?